File: bindings/jvm/src/main/java/io/vinarytree/libdictenstein/Dictionary.java
package io.vinarytree.libdictenstein;

import static java.lang.foreign.ValueLayout.ADDRESS;
import static java.lang.foreign.ValueLayout.JAVA_BYTE;
import static java.lang.foreign.ValueLayout.JAVA_LONG;

import io.vinarytree.interop.DictionaryResource;
import io.vinarytree.interop.DictionaryBatchLimits;
import io.vinarytree.interop.DictionaryEntry;
import io.vinarytree.interop.DictionaryEntryIterator;
import io.vinarytree.interop.DictionaryKey;
import io.vinarytree.interop.DictionarySnapshot;
import io.vinarytree.interop.InteropLayouts;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.ref.Cleaner;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.OptionalLong;
import java.util.Iterator;
import java.util.stream.Stream;

/** Common read-only surface for project-owned native dictionaries. */
public abstract class Dictionary implements DictionaryResource, Iterable<DictionaryEntry> {
    private static final Cleaner CLEANER = Cleaner.create();

    private static final class State implements Runnable {
        private MemorySegment handle;
        private final Arena arena;

        State(MemorySegment handle, Arena arena) {
            this.handle = handle;
            this.arena = arena;
        }

        synchronized MemorySegment handle() {
            if (handle.equals(MemorySegment.NULL)) {
                throw new IllegalStateException("dictionary is closed");
            }
            return handle;
        }

        @Override public synchronized void run() {
            if (!handle.equals(MemorySegment.NULL)) {
                Native.free(handle);
                handle = MemorySegment.NULL;
                arena.close();
            }
        }
    }

    /** Lookup result preserving absent, valueless, and valued terms. */
    public record Lookup(boolean present, OptionalLong value) {
        public Lookup {
            Objects.requireNonNull(value, "value");
            if (!present && value.isPresent()) {
                throw new IllegalArgumentException("an absent term cannot have a value");
            }
        }
    }

    /** Raw byte term for one batched insert/update. */
    public record TextEntry(byte[] term, OptionalLong value) {
        public TextEntry {
            term = Objects.requireNonNull(term, "term").clone();
            Objects.requireNonNull(value, "value");
        }

        @Override public byte[] term() { return term.clone(); }
    }

    /** U64-token term for one batched insert/update. */
    public record U64Entry(long[] term, OptionalLong value) {
        public U64Entry {
            term = Objects.requireNonNull(term, "term").clone();
            Objects.requireNonNull(value, "value");
        }

        @Override public long[] term() { return term.clone(); }
    }

    private final UnitDomain domain;
    private final State state;
    private final Cleaner.Cleanable cleanable;
    private final MemorySegment resource;

    Dictionary(UnitDomain domain, MemorySegment handle) {
        this.domain = Objects.requireNonNull(domain, "domain");
        Arena owned = Arena.ofShared();
        state = new State(Objects.requireNonNull(handle, "handle"), owned);
        try {
            resource = owned.allocate(InteropLayouts.RESOURCE);
            Native.check(Native.resource(handle, resource));
            cleanable = CLEANER.register(this, state);
        } catch (Throwable failure) {
            state.run();
            throw failure;
        }
    }

    /** Native ABI version (LDICT_ABI_VERSION); always 1 for this family. */
    public static int abiVersion() { return Native.abiVersion(); }

    /** Compatible-additions revision within the ABI version (LDICT_API_REVISION). */
    public static int apiRevision() { return Native.apiRevision(); }

    /** Unit domain fixed at construction/open time. */
    public final UnitDomain domain() { return domain; }

    /** Stable native backend identifier. */
    public final int kind() {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(java.lang.foreign.ValueLayout.JAVA_INT);
            Native.check(Native.kind(handle(), output));
            return output.get(java.lang.foreign.ValueLayout.JAVA_INT, 0);
        }
    }

    /** Native capability bitset. */
    public final long capabilities() {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_LONG);
            Native.check(Native.capabilities(handle(), output));
            return output.get(JAVA_LONG, 0);
        }
    }

    /** Number of visible terms. */
    public final long size() {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_LONG);
            Native.check(Native.len(handle(), output));
            return output.get(JAVA_LONG, 0);
        }
    }

    /**
     * Materialize one immutable, repeatable lexicographic revision.
     *
     * <p>Every key is copied once out of its borrowed native batch. The returned collection owns
     * all of its storage and remains valid after this dictionary is changed or closed.
     */
    public final DictionarySnapshot snapshot() { return entriesSnapshot(); }

    /** Materialized immutable entry-list view of one captured revision. */
    public final List<DictionaryEntry> entries() { return snapshot().orderedEntries(); }

    /** Materialized immutable key-set view of one captured revision. */
    public final java.util.Set<DictionaryKey> keys() { return snapshot().keys(); }

    /** Materialized immutable value-collection view of one captured revision. */
    public final java.util.Collection<java.util.Optional<io.vinarytree.interop.UnsignedLong>> values() {
        return snapshot().entries().values();
    }

    /** Materialized immutable map view of one captured revision. */
    public final Map<DictionaryKey, java.util.Optional<io.vinarytree.interop.UnsignedLong>> asMap() {
        return snapshot().entries();
    }

    /**
     * Materialize an independent mutable DynamicDAWG from two immutable input revisions.
     *
     * <p>The native implementation performs one linear ordered merge and freezes the result once;
     * no Java-side entry materialization or per-entry FFM call occurs.
     */
    public final DynamicDawg algebra(
            Dictionary right, AlgebraOperation operation, ValueMerge valueMerge) {
        Objects.requireNonNull(right, "right");
        Objects.requireNonNull(operation, "operation");
        Objects.requireNonNull(valueMerge, "valueMerge");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(ADDRESS);
            MemorySegment result = newHandle(Native.algebra(
                    handle(), right.handle(), operation.nativeValue, valueMerge.nativeValue, output),
                    output);
            return DynamicDawg.adopt(domain, result);
        }
    }

    /** Return keys in either input; duplicate values default to right-biased selection. */
    public final DynamicDawg union(Dictionary right) { return union(right, ValueMerge.LAST); }

    /** Return keys in either input with the requested duplicate-value policy. */
    public final DynamicDawg union(Dictionary right, ValueMerge valueMerge) {
        return algebra(right, AlgebraOperation.UNION, valueMerge);
    }

    /** Return shared keys; duplicate values default to the optional-u64 lattice meet. */
    public final DynamicDawg intersection(Dictionary right) {
        return intersection(right, ValueMerge.LATTICE_MEET);
    }

    /** Return shared keys with the requested duplicate-value policy. */
    public final DynamicDawg intersection(Dictionary right, ValueMerge valueMerge) {
        return algebra(right, AlgebraOperation.INTERSECTION, valueMerge);
    }

    /** Return keys present in this dictionary but absent from {@code right}. */
    public final DynamicDawg difference(Dictionary right) {
        return algebra(right, AlgebraOperation.DIFFERENCE, ValueMerge.FIRST);
    }

    /** Return keys present in exactly one input. */
    public final DynamicDawg symmetricDifference(Dictionary right) {
        return algebra(right, AlgebraOperation.SYMMETRIC_DIFFERENCE, ValueMerge.FIRST);
    }

    /** Open a single-pass native-backed cursor with a default batch bound of 256 entries. */
    public final DictionaryEntryIterator openEntryStream() { return entryIterator(); }

    /** Open a single-pass native-backed cursor with the requested positive entry batch bound. */
    public final DictionaryEntryIterator openEntryStream(int batchSize) {
        return entryIterator(batchLimits(batchSize));
    }

    /** Open a single-pass cursor with explicit native descriptor/arena limits. */
    public final DictionaryEntryIterator openEntryStream(DictionaryBatchLimits limits) {
        return entryIterator(Objects.requireNonNull(limits, "limits"));
    }

    /**
     * Open a sequential Java stream over one captured revision.
     *
     * <p>Exhaustion closes the cursor automatically. Use try-with-resources when a pipeline may
     * short-circuit so {@link Stream#close()} cancels the cursor promptly.
     */
    public final Stream<DictionaryEntry> streamEntries() { return entryStream(); }

    /** Open a sequential Java stream with the requested positive entry batch bound. */
    public final Stream<DictionaryEntry> streamEntries(int batchSize) {
        return entryStream(batchLimits(batchSize));
    }

    /** Open a sequential Java stream with explicit native descriptor/arena limits. */
    public final Stream<DictionaryEntry> streamEntries(DictionaryBatchLimits limits) {
        return entryStream(Objects.requireNonNull(limits, "limits"));
    }

    /**
     * Return a repeatable iterator backed by an owned materialized snapshot.
     * Use {@link #openEntryStream(int)} for bounded-memory traversal.
     */
    @Override public final Iterator<DictionaryEntry> iterator() { return snapshot().iterator(); }

    private static DictionaryBatchLimits batchLimits(int batchSize) {
        if (batchSize <= 0) throw new IllegalArgumentException("batchSize must be positive");
        long units = Math.max(65_536L, Math.multiplyExact((long) batchSize, 64L));
        return new DictionaryBatchLimits(batchSize, units, batchSize);
    }

    /** Test a Unicode term. */
    public final boolean contains(String term) {
        return containsText(Objects.requireNonNull(term, "term").getBytes(StandardCharsets.UTF_8));
    }

    /** Test a raw byte term. */
    public final boolean contains(byte[] term) {
        return containsText(Objects.requireNonNull(term, "term"));
    }

    /** Test a u64-token term. */
    public final boolean contains(long[] term) {
        ensureDomain(UnitDomain.U64);
        Objects.requireNonNull(term, "term");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.containsU64(handle(), longs(arena, term), term.length, output));
            return booleanOutput(output);
        }
    }

    /** Read a Unicode term. */
    public final Lookup get(String term) {
        return getText(Objects.requireNonNull(term, "term").getBytes(StandardCharsets.UTF_8));
    }

    /** Read a raw byte term. */
    public final Lookup get(byte[] term) {
        return getText(Objects.requireNonNull(term, "term"));
    }

    /** Read a u64-token term. */
    public final Lookup get(long[] term) {
        ensureDomain(UnitDomain.U64);
        Objects.requireNonNull(term, "term");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment found = arena.allocate(JAVA_BYTE);
            MemorySegment value = arena.allocate(Native.OPTIONAL_U64);
            Native.check(Native.getU64(handle(), longs(arena, term), term.length, found, value));
            return lookup(found, value);
        }
    }

    @Override public final MemorySegment resourceSegment() {
        handle();
        return resource;
    }

    @Override public final void close() { cleanable.clean(); }

    protected final MemorySegment handle() { return state.handle(); }

    protected final boolean insert(String term, OptionalLong value) {
        return insert(Objects.requireNonNull(term, "term").getBytes(StandardCharsets.UTF_8), value);
    }

    protected final boolean insert(byte[] term, OptionalLong value) {
        ensureTextDomain();
        Objects.requireNonNull(term, "term");
        Objects.requireNonNull(value, "value");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.insertText(handle(), bytes(arena, term), term.length,
                    optional(arena, value), output));
            return booleanOutput(output);
        }
    }

    protected final boolean insert(long[] term, OptionalLong value) {
        ensureDomain(UnitDomain.U64);
        Objects.requireNonNull(term, "term");
        Objects.requireNonNull(value, "value");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.insertU64(handle(), longs(arena, term), term.length,
                    optional(arena, value), output));
            return booleanOutput(output);
        }
    }

    protected final boolean removeTerm(String term) {
        return removeTerm(Objects.requireNonNull(term, "term").getBytes(StandardCharsets.UTF_8));
    }

    protected final boolean removeTerm(byte[] term) {
        ensureTextDomain();
        Objects.requireNonNull(term, "term");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.removeText(handle(), bytes(arena, term), term.length, output));
            return booleanOutput(output);
        }
    }

    protected final boolean removeTerm(long[] term) {
        ensureDomain(UnitDomain.U64);
        Objects.requireNonNull(term, "term");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.removeU64(handle(), longs(arena, term), term.length, output));
            return booleanOutput(output);
        }
    }

    protected final long insertAllStrings(Map<String, OptionalLong> entries) {
        Objects.requireNonNull(entries, "entries");
        return insertAllBytes(entries.entrySet().stream()
                .map(entry -> new TextEntry(
                        Objects.requireNonNull(entry.getKey(), "term")
                                .getBytes(StandardCharsets.UTF_8),
                        Objects.requireNonNull(entry.getValue(), "value")))
                .toList());
    }

    protected final long insertAllBytes(List<TextEntry> entries) {
        ensureTextDomain();
        Objects.requireNonNull(entries, "entries");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment nativeEntries = arena.allocate(Native.ENTRY, Math.max(1, entries.size()));
            for (int index = 0; index < entries.size(); index++) {
                TextEntry entry = Objects.requireNonNull(entries.get(index), "entry");
                byte[] term = entry.term;
                writeEntry(nativeEntries.asSlice(index * Native.ENTRY.byteSize(),
                        Native.ENTRY.byteSize()), bytes(arena, term), term.length,
                        entry.value, arena);
            }
            MemorySegment output = arena.allocate(JAVA_LONG);
            Native.check(Native.insertTextBatch(handle(), nativeEntries, entries.size(), output));
            return output.get(JAVA_LONG, 0);
        }
    }

    protected final long insertAllU64(List<U64Entry> entries) {
        ensureDomain(UnitDomain.U64);
        Objects.requireNonNull(entries, "entries");
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment nativeEntries = arena.allocate(Native.ENTRY, Math.max(1, entries.size()));
            for (int index = 0; index < entries.size(); index++) {
                U64Entry entry = Objects.requireNonNull(entries.get(index), "entry");
                long[] term = entry.term;
                writeEntry(nativeEntries.asSlice(index * Native.ENTRY.byteSize(),
                        Native.ENTRY.byteSize()), longs(arena, term), term.length,
                        entry.value, arena);
            }
            MemorySegment output = arena.allocate(JAVA_LONG);
            Native.check(Native.insertU64Batch(handle(), nativeEntries, entries.size(), output));
            return output.get(JAVA_LONG, 0);
        }
    }

    protected final void clearTerms() { Native.check(Native.clear(handle())); }

    protected final long compactTerms() {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_LONG);
            Native.check(Native.compact(handle(), output));
            return output.get(JAVA_LONG, 0);
        }
    }

    static MemorySegment newHandle(int status, MemorySegment output) {
        Native.check(status);
        MemorySegment handle = output.get(ADDRESS, 0);
        if (handle.equals(MemorySegment.NULL)) {
            throw new IllegalStateException("native constructor returned a null handle");
        }
        return handle;
    }

    static MemorySegment bytes(Arena arena, byte[] value) {
        MemorySegment result = arena.allocate(Math.max(1, value.length), 1);
        if (value.length != 0) result.asSlice(0, value.length).copyFrom(MemorySegment.ofArray(value));
        return result;
    }

    static MemorySegment longs(Arena arena, long[] value) {
        MemorySegment result = arena.allocate(
                Math.max(1, value.length) * JAVA_LONG.byteSize(), JAVA_LONG.byteAlignment());
        for (int index = 0; index < value.length; index++) result.setAtIndex(JAVA_LONG, index, value[index]);
        return result;
    }

    static MemorySegment optional(Arena arena, OptionalLong value) {
        MemorySegment result = arena.allocate(Native.OPTIONAL_U64);
        result.fill((byte) 0);
        result.set(JAVA_LONG, Native.OPTIONAL_VALUE, value.orElse(0));
        result.set(JAVA_BYTE, Native.OPTIONAL_PRESENT, (byte) (value.isPresent() ? 1 : 0));
        return result;
    }

    static void writeEntry(MemorySegment entry, MemorySegment data, long length,
                           OptionalLong value, Arena arena) {
        entry.set(ADDRESS, Native.ENTRY_DATA, data);
        entry.set(JAVA_LONG, Native.ENTRY_LEN, length);
        entry.asSlice(Native.ENTRY_VALUE, Native.OPTIONAL_U64.byteSize())
                .copyFrom(optional(arena, value));
    }

    private boolean containsText(byte[] term) {
        ensureTextDomain();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment output = arena.allocate(JAVA_BYTE);
            Native.check(Native.containsText(handle(), bytes(arena, term), term.length, output));
            return booleanOutput(output);
        }
    }

    private Lookup getText(byte[] term) {
        ensureTextDomain();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment found = arena.allocate(JAVA_BYTE);
            MemorySegment value = arena.allocate(Native.OPTIONAL_U64);
            Native.check(Native.getText(handle(), bytes(arena, term), term.length, found, value));
            return lookup(found, value);
        }
    }

    private static Lookup lookup(MemorySegment found, MemorySegment value) {
        boolean present = booleanOutput(found);
        OptionalLong optional = value.get(JAVA_BYTE, Native.OPTIONAL_PRESENT) == 0
                ? OptionalLong.empty()
                : OptionalLong.of(value.get(JAVA_LONG, Native.OPTIONAL_VALUE));
        return new Lookup(present, optional);
    }

    private static boolean booleanOutput(MemorySegment output) {
        return output.get(JAVA_BYTE, 0) != 0;
    }

    private void ensureTextDomain() {
        if (domain == UnitDomain.U64) {
            throw new IllegalStateException("u64 dictionary requires long[] terms");
        }
        handle();
    }

    private void ensureDomain(UnitDomain expected) {
        if (domain != expected) throw new IllegalStateException("dictionary unit domain is " + domain);
        handle();
    }
}

File: bindings/dotnet/src/VinaryTree.Libdictenstein/Dictionary.cs
using System.Text;
using VinaryTree.Interop;

namespace VinaryTree.Libdictenstein;

/// <summary>Transition unit domain.</summary>
public enum UnitDomain : uint
{
    /// <summary>Raw bytes.</summary>
    Byte = 1,
    /// <summary>Unicode scalar values.</summary>
    UnicodeScalar = 2,
    /// <summary>Unsigned 64-bit tokens.</summary>
    U64 = 3,
}

/// <summary>Concrete dictionary implementation.</summary>
public enum BackendKind : uint
{
    /// <summary>Mutable dynamic directed acyclic word graph.</summary>
    DynamicDawg = 1,
    /// <summary>Immutable double-array trie.</summary>
    DoubleArrayTrie = 2,
    /// <summary>Suffix-compressed directed acyclic word graph.</summary>
    Scdawg = 3,
    /// <summary>Filesystem-backed adaptive radix trie.</summary>
    PersistentArtrie = 4,
    /// <summary>Bidirectional persistent term/index vocabulary.</summary>
    PersistentVocabulary = 5,
}

/// <summary>Exact key-set operation evaluated over immutable input revisions.</summary>
public enum AlgebraOperation : uint
{
    /// <summary>Keys present in either input.</summary>
    Union = 1,
    /// <summary>Keys present in both inputs.</summary>
    Intersection = 2,
    /// <summary>Keys present in the left input but not the right.</summary>
    Difference = 3,
    /// <summary>Keys present in exactly one input.</summary>
    SymmetricDifference = 4,
}

/// <summary>Conflict policy for optional u64 values on keys in both inputs.</summary>
public enum ValueMerge : uint
{
    /// <summary>Keep the left value.</summary>
    First = 1,
    /// <summary>Keep the right value.</summary>
    Last = 2,
    /// <summary>Use the optional-u64 lattice join.</summary>
    LatticeJoin = 3,
    /// <summary>Use the optional-u64 lattice meet.</summary>
    LatticeMeet = 4,
}

/// <summary>Distinguishes absent terms, present terms without values, and mapped terms.</summary>
public readonly record struct Lookup(bool Found, ulong? Value);

/// <summary>Common exact-lookup and resource API for all backends.</summary>
public abstract class Dictionary : IDictionaryResource, IDisposable,
    IEnumerable<VinaryTree.Interop.DictionaryEntry>
{
    internal DictionaryHandle Handle { get; }
    internal Dictionary(nint handle) => Handle = handle != 0 ? new DictionaryHandle(handle) : throw new InvalidOperationException("native dictionary was not returned");

    /// <summary>Native ABI version (LDICT_ABI_VERSION); always 1 for this family.</summary>
    public static uint AbiVersion => Native.AbiVersion();
    /// <summary>Compatible-additions revision within the ABI version (LDICT_API_REVISION).</summary>
    public static uint ApiRevision => Native.ApiRevision();

    /// <summary>The concrete backend.</summary>
    public BackendKind Kind { get { Native.Check(Native.Kind(Handle, out uint result)); return (BackendKind)result; } }
    /// <summary>The stable capability bit set.</summary>
    public ulong Capabilities { get { Native.Check(Native.Capabilities(Handle, out ulong result)); return result; } }
    /// <summary>Number of complete terms.</summary>
    public nuint Count { get { Native.Check(Native.Length(Handle, out nuint result)); return result; } }

    /// <summary>
    /// Materialize one immutable, repeatable lexicographic revision. Keys are copied once out of
    /// borrowed native batches and remain valid after this dictionary changes or is disposed.
    /// </summary>
    public VinaryTree.Interop.DictionarySnapshot Snapshot() => this.SnapshotEntries();

    /// <summary>Materialized immutable entry collection for one revision.</summary>
    public IReadOnlyCollection<VinaryTree.Interop.DictionaryEntry> Entries => Snapshot();
    /// <summary>Materialized immutable dictionary view for one revision.</summary>
    public IReadOnlyDictionary<VinaryTree.Interop.DictionaryKey, ulong?> AsReadOnlyDictionary() => Snapshot().Entries;
    /// <summary>Open a single-pass native-backed cursor with the requested batch bound.</summary>
    public DictionaryEntryEnumerator OpenEntryStream(int batchSize = 256) =>
        this.OpenEntryEnumerator(BatchLimits(batchSize));
    /// <summary>Open a one-shot enumerable stream with the requested batch bound.</summary>
    public DictionaryEntryStream StreamEntries(int batchSize = 256) =>
        VinaryTree.Interop.DictionaryCollectionExtensions.StreamEntries(this, BatchLimits(batchSize));

    /// <summary>
    /// Materialize an independent mutable DynamicDAWG with one native ordered merge.
    /// </summary>
    public DynamicDawg Algebra(Dictionary right, AlgebraOperation operation,
        ValueMerge valueMerge = ValueMerge.Last)
    {
        ArgumentNullException.ThrowIfNull(right);
        Native.Check(Native.Algebra(Handle, right.Handle, (uint)operation, (uint)valueMerge,
            out nint result));
        return DynamicDawg.Adopt(result);
    }

    /// <summary>Return keys in either input with right-biased values by default.</summary>
    public DynamicDawg Union(Dictionary right, ValueMerge valueMerge = ValueMerge.Last) =>
        Algebra(right, AlgebraOperation.Union, valueMerge);

    /// <summary>Return shared keys with lattice-meet values by default.</summary>
    public DynamicDawg Intersection(Dictionary right,
        ValueMerge valueMerge = ValueMerge.LatticeMeet) =>
        Algebra(right, AlgebraOperation.Intersection, valueMerge);

    /// <summary>Return keys present only in this dictionary.</summary>
    public DynamicDawg Difference(Dictionary right) =>
        Algebra(right, AlgebraOperation.Difference, ValueMerge.First);

    /// <summary>Return keys present in exactly one input.</summary>
    public DynamicDawg SymmetricDifference(Dictionary right) =>
        Algebra(right, AlgebraOperation.SymmetricDifference, ValueMerge.First);

    /// <summary>Set-union operator using right-biased duplicate values.</summary>
    public static DynamicDawg operator |(Dictionary left, Dictionary right) => left.Union(right);
    /// <summary>Set-intersection operator using lattice-meet duplicate values.</summary>
    public static DynamicDawg operator &(Dictionary left, Dictionary right) => left.Intersection(right);
    /// <summary>Set-difference operator.</summary>
    public static DynamicDawg operator -(Dictionary left, Dictionary right) => left.Difference(right);
    /// <summary>Symmetric-set-difference operator.</summary>
    public static DynamicDawg operator ^(Dictionary left, Dictionary right) => left.SymmetricDifference(right);
    /// <summary>
    /// Return a native-backed snapshot enumerator. C# foreach disposes it; manually controlled
    /// enumeration must also dispose it when stopping early.
    /// </summary>
    public IEnumerator<VinaryTree.Interop.DictionaryEntry> GetEnumerator() => OpenEntryStream();
    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();

    private static DictionaryBatchLimits BatchLimits(int batchSize)
    {
        if (batchSize <= 0) throw new ArgumentOutOfRangeException(nameof(batchSize));
        nuint units = checked((nuint)Math.Max(65_536L, checked((long)batchSize * 64)));
        return new((nuint)batchSize, units, (nuint)batchSize);
    }

    /// <summary>Test a Unicode or raw-byte term.</summary>
    public unsafe bool Contains(string term)
    {
        byte[] text = Encoding.UTF8.GetBytes(term);
        fixed (byte* data = text) { Native.Check(Native.ContainsText(Handle, data, (nuint)text.Length, out byte result)); return result != 0; }
    }
    /// <summary>Look up a Unicode or raw-byte term.</summary>
    public unsafe Lookup Get(string term)
    {
        byte[] text = Encoding.UTF8.GetBytes(term);
        fixed (byte* data = text)
        {
            Native.Check(Native.GetText(Handle, data, (nuint)text.Length, out byte found, out OptionalU64 value));
            return new(found != 0, value.HasValue != 0 ? value.Value : null);
        }
    }
    /// <summary>Test a u64-token term.</summary>
    public unsafe bool Contains(ReadOnlySpan<ulong> term)
    {
        fixed (ulong* data = term) { Native.Check(Native.ContainsU64(Handle, data, (nuint)term.Length, out byte result)); return result != 0; }
    }
    /// <summary>Look up a u64-token term.</summary>
    public unsafe Lookup Get(ReadOnlySpan<ulong> term)
    {
        fixed (ulong* data = term)
        {
            Native.Check(Native.GetU64(Handle, data, (nuint)term.Length, out byte found, out OptionalU64 value));
            return new(found != 0, value.HasValue != 0 ? value.Value : null);
        }
    }

    /// <inheritdoc />
    public unsafe TResult WithResource<TResult>(ResourceCallback<TResult> callback)
    {
        ArgumentNullException.ThrowIfNull(callback);
        bool retained = false;
        try
        {
            Handle.DangerousAddRef(ref retained);
            NativeResource resource = default;
            Native.Check(Native.Resource(Handle, &resource));
            return callback(&resource);
        }
        finally { if (retained) Handle.DangerousRelease(); }
    }

    internal static unsafe bool InsertText(DictionaryHandle handle, string term, ulong? value)
    {
        byte[] text = Encoding.UTF8.GetBytes(term);
        fixed (byte* data = text) { Native.Check(Native.InsertText(handle, data, (nuint)text.Length, OptionalU64.From(value), out byte inserted)); return inserted != 0; }
    }
    internal static unsafe bool RemoveText(DictionaryHandle handle, string term)
    {
        byte[] text = Encoding.UTF8.GetBytes(term);
        fixed (byte* data = text) { Native.Check(Native.RemoveText(handle, data, (nuint)text.Length, out byte removed)); return removed != 0; }
    }
    internal static unsafe bool InsertTokens(DictionaryHandle handle, ReadOnlySpan<ulong> term, ulong? value)
    {
        fixed (ulong* data = term) { Native.Check(Native.InsertU64(handle, data, (nuint)term.Length, OptionalU64.From(value), out byte inserted)); return inserted != 0; }
    }
    internal static unsafe bool RemoveTokens(DictionaryHandle handle, ReadOnlySpan<ulong> term)
    {
        fixed (ulong* data = term) { Native.Check(Native.RemoveU64(handle, data, (nuint)term.Length, out byte removed)); return removed != 0; }
    }

    /// <inheritdoc />
    public void Dispose() => Handle.Dispose();
}

internal static class TextBatch
{
    internal unsafe delegate TResult DescriptorAction<TResult>(TextEntry* entries, nuint count);

    internal static unsafe TResult WithDescriptors<TResult>(IReadOnlyCollection<KeyValuePair<string, ulong?>> entries, DescriptorAction<TResult> action)
    {
        var encoded = new byte[entries.Count][];
        int total = 0;
        int encodedIndex = 0;
        foreach (KeyValuePair<string, ulong?> entry in entries)
        {
            encoded[encodedIndex] = Encoding.UTF8.GetBytes(entry.Key);
            total = checked(total + encoded[encodedIndex].Length);
            encodedIndex++;
        }
        byte[] arena = new byte[Math.Max(1, total)];
        var descriptors = new TextEntry[entries.Count];
        fixed (byte* basePointer = arena)
        fixed (TextEntry* descriptorPointer = descriptors)
        {
            int offset = 0;
            int index = 0;
            foreach (KeyValuePair<string, ulong?> entry in entries)
            {
                encoded[index].CopyTo(arena, offset);
                descriptors[index] = new TextEntry { Data = basePointer + offset, Length = (nuint)encoded[index].Length, Value = OptionalU64.From(entry.Value) };
                offset += encoded[index].Length;
                index++;
            }
            return action(descriptorPointer, (nuint)descriptors.Length);
        }
    }
}
